api/enki.py

```diff
@@ -249,17 +294,136 @@ def checkout(self, patron, pin, licensepool, internal_format):
         )
         return loan
 
-    def get_loan(barcode, pin, book_id):
-        self.log.debug ("Sending checkout request for %d" % book_id)
-        now = datetime.datetime.utcnow()
+    def loan_request(self, barcode, pin, book_id):
+        self.log.debug ("Sending checkout request for %s" % book_id)
         url = str(self.base_url) + str(self.user_endpoint)
         args = dict()
         args['method'] = "getSELink"
-        args['username'] = barcide
+        args['username'] = barcode
         args['password'] = pin
         args['lib'] = self.library_id
         args['id'] = book_id
+
+        response = self.request(url, method='get', params=args)
+        return response
+
+    def fulfill(self, patron, pin, licensepool, internal_format):
+        book_id = licensepool.identifier.identifier
+        response = self.loan_request(patron.authorization_identifier, pin, book_id)
+        if response.status_code != 200:
+            raise CannotFulfill(response.status_code)
+        result = json.loads(response.content)['result']
+        if not result['success']:
+            message = result['message']
+            if "That record is already checked out to you" in message:
+                # this might be okay
+                pass
+            elif "There are no available copies" in message:
+                # TODO: raise no copies error
+                self.log.error("There are no copies of this book available.")
+                raise CirculationException()
+            elif "Login unsuccessful" in message:
+                # TODO: raise invalid barcode/password error
+                self.log.error("User validation against Enki server was unsuccessful.")
+                raise CirculationException()
+        media_type = self.get_enki_media_type(book_id)
+        url, item_type, expires = self.parse_fulfill_result(result)
+
+        if not media_type and item_type == 'epub':
+            media_type = self.epub
```

Comment:
Just want to confirm that the URL here leads to an EPUB and not an ACSM file.

Reply:
My reasoning here is that we should always get either "free" or "acs" for media_type, and in either of those cases we'll have the right type string to return in FulfillmentInfo. If we get something unexpected there (either because we don't have all the information now or because something changes later) we'll assume it's an EPUB and handle it then if that breaks things.